L67-cpp-callback-api.md

```diff
@@ -0,0 +1,105 @@
+
+Provide an asynchronous gRPC API for C++ in which the completion of RPC actions in the library will result in callbacks to user code,
+
+## Background
+
+Since its initial release, gRPC has provided two C++ APIs:
+
+* Synchronous API
+   - All RPC actions (such as unary calls, streaming reads, streaming writes, etc.) block for completion
+* Completion-queue-based asynchronous API
+   - Application associates each RPC action that it initiates with a tag
+   - The library performs each RPC action
+   - The library posts the tag of a completed action onto a completion queue
+   - The application must poll the completion queue to determine which asynchronously-initiated actions have completed
+
+The goal of the synchronous version is to be easy to program. However, this comes at the cost of high thread-switching overhead and high thread storage for systems with many concurrent RPCs. On the other hand, the asynchronous API allows the application full control over its threading and thus can scale further. It does, however, present challenges in deciding how many completion queues to use and how many threads to use for polling them, as one can either optimize for reducing thread hops, avoiding stranding, reducing CQ contention, or improving locality. These goals are often in conflict and require substantial tuning.
+
+### Related proposals
+
+* The [EventManager](https://www.github.com/grpc/proposal/XXX) is critical to the operation of the callback API.
+* The C++ callback API is built on top of a new [callback completion queue in core](https://github.com/grpc/proposal/pull/181)
+* The API structure has substantial similarities to the gRPC-Node and gRPC-Java APIs
+
+## Proposal
+
+The callback API is designed to have the performance and thread scalability of an asynchronous API without the burdensome programming model of the completion-queue-based model. In particular, the following are fundamental guiding principles of the API:
+
+* Library directly calls user-specified code at the completion of RPC actions.
+* No explicit polling required for notification of completion of RPC actions.
+   - In practice, these requirements mean that there must be a library-controlled poller for monitoring such actions. This _EventManager_ is described in another gRFC.
+* As in the synchronous API, server RPCs have an application-defined method handler function as part of their service definition. The library invokes this method handler when a new server RPC starts.
+* LIke the synchronnous API and unlike the completion-queue-based asynchronous API, there is no need for the application to "request" new server RPCs. Server RPC context structures will be allocated and have their resources allocated as and when RPCs arrive at the server.
+
+### Reactor model
+
+The most general form of the callback API is built around a _reactor_ model. Each type of RPC has a reactor base class provided by the library. These types are:
+
+* `ClientUnaryReactor` for unary calls
```

Comment:
This could be a table, if you're feeling ambitious. :)

Reply:
I think the API is the limit of my ambition, formatting not so much....